### ISSUE TITLE
Minetest serverthread 100% busy when pipes are busy / leaking

### ISSUE THREAD

**Issue by tchncs** [minetest-mods/pipeworks#140]
Hi there, i just figured out (with help from Fixer in the irc), that this mod creates huge laggs somehow.
I dont really know whats changed, its with the current state and with a commit from before, an older Minetest dev state and the current one.

Performance Info:
https://haste.tchncs.de/dewohufazo.apache

I dont really know the indicator, but its unplayable now.
Well we switched to a [forked farming mod](https://git.tchncs.de/Illuna-Minetest/farming) but it also happens without this. 
Regards

**Comment by tchncs** [comment]
Someone forgot to configure his teleport tube so all his farmed stuff was compressed in the pipes!
This problem is solved!

**Comment by sofar** [comment]
Maybe this is something that can be fixed, or prevented somehow.

**Comment by HybridDog** [comment]
sofar, l'm not sure in which way here, maybe the function_delayer mod can be used to queue the pipeworks actions if the server is lagging

tchncs, how did you make this performance info?

**Comment by tchncs** [comment]
@HybridDog i added the following into my minetest.conf:

```
mod_profiling = true
detailed_profiling = true
```
This adds the `/save_mod_profile` command wich smashes the table into the logfile.

**Comment by HybridDog** [comment]
thanks, l didn't know about this command

**Comment by tchncs** [comment]
All thanks to @fixer-007 :cake:

**Comment by HybridDog** [comment]
l've got a problem:
lf l enable mod_profiling and then drop an item, minetest crashes:
```
2016-05-23 22:05:35: ERROR[Main]: ServerError: Lua: Runtime error from mod 'pipeworks' in callback luaentity_Step(): /home/cilla/.minetest/mods/pipeworks/vacuum_tubes.lua:250: attempt to perform arithmetic on local 'timer' (a nil value)
2016-05-23 22:05:35: ERROR[Main]: stack traceback:
2016-05-23 22:05:35: ERROR[Main]: 	/home/cilla/.minetest/mods/pipeworks/vacuum_tubes.lua:250: in function </home/cilla/.minetest/mods/pipeworks/vacuum_tubes.lua:248>
2016-05-23 22:05:35: ERROR[Main]: 	/usr/share/minetest/builtin/game/mod_profiling.lua:279: in function </usr/share/minetest/builtin/game/mod_profiling.lua:277>
diplazer GGun&Tp is cleaned (0 0)
2016-05-23 22:05:35: ACTION[Server]: singleplayer leaves game. List of players: 
AL lib: FreeContext: (0x1ba0630) Deleting 2 Source(s)
2016-05-23 22:05:35: ACTION[Main]: [ltool] Tree data saved to /home/cilla/.minetest/worlds/testworld/ltool.mt.
[biome_lib] Stand by, playing out the rest of the aircheck mapblock log
(there are 0 entries)...
[biome_lib] Stand by, playing out the rest of the no-aircheck mapblock log
(there are 0 entries)...
terminate called after throwing an instance of 'LuaError'
  what():  Runtime error from mod 'pipeworks' in callback luaentity_GetStaticdata(): stack overflow
stack traceback:
	[C]: in function 'get_us_time'
	/usr/share/minetest/builtin/game/mod_profiling.lua:311: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	...
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
	/usr/share/minetest/builtin/game/mod_profiling.lua:312: in function </usr/share/minetest/builtin/game/mod_profiling.lua:310>
Aborted
```
for any reason now it doesn't tell the stack overflow error
before it printed that it crashes in the on_activate

my pipeworks version overrides the on_activate of dropped entities, l don't know why it stacks the function
https://github.com/HybridDog/pipeworks/blob/hversion/vacuum_tubes.lua#L243

**Comment by Fixer-007** [comment]
Why close? I agree with sofar, maybe there must be some function to control this.

**Comment by sofar** [comment]
I don't understand the root cause just yet, but if too many items are getting pumped to a certain location, it should be at least detectable.

**Comment by VanessaE** [comment]
This problem has since been solved by the "break tubes" feature (33c4c9d552c7f113a21405456e67c0e4752674a8 to f3a74bb618d96a599e2cc0f416c5c67320181275, et al.), and pipeworks can skip globalsteps if dtime is too high.
